File: sorting/sorting.js
/*
//bubble sort
const bubbleSort = function (arr) {
	for (let i = 0; i < arr.length; i += 1) {
		for (let j = i; j < arr.length; j += 1) {
			if (arr[i] > arr[j]) {
				let temp = arr[i];
				arr[i] = arr[j];
				arr[j] = temp;
			}
		}
	}
}

//insertion sort
const insertionSort = function (arr) {
	for (let i = 1; i < arr.length; i += 1) {
		let min = i;
		const minValue = arr[i];
		for (let j = i - 1; j >= 0; j -= 1) {
			if (arr[j] > minValue) {
				arr[j+1] = arr[j];
				min = j;
			}
		}
		arr[min] = minValue;
	}
}

//selection sort
const selectionSort = function (arr) {
	for (let i = 0; i < arr.length; i += 1) {
		let min = i;
		for (let j = i; j < arr.length; j += 1) {
			if (arr[j] < arr[min]) min = j;
		}
		let temp = arr[i];
		arr[i] = arr[min];
		arr[min] = temp;
	}
}

//count sort
const countSort = function (arr) {
	const max = Math.max(...arr);
	let bucket = new Array(max + 1);
	for (let i = 0; i < bucket.length; i += 1) {
		bucket[i] = 0;
	}
	for (let i = 0; i < arr.length; i += 1) {
		bucket[arr[i]] += 1;
	}
	let i = 0;
	let sortedArr = [];
	while (i < bucket.length) {
		if (bucket[i] > 0) {
			sortedArr.push(i);
			bucket[i] -= 1;
		} else i += 1;
	}
	return sortedArr;
}


//bucket sort
const bucketSort = function (arr) {
	let bucket = [];
	for (let i = 0; i <= 9; i += 1) {
		bucket.push([]);
	}
	for (let i = 0; i < arr.length; i += 1) {
		bucket[Math.floor(arr[i] / 10)].push(arr[i]);
	}
	for (let i = 0; i < bucket.length; i += 1) {
		bucket[i].sort((a, b) => a - b);
	}
	let sortedArr = [];
	for (let i = 0; i < bucket.length; i += 1) {
		if (bucket[i].length > 0) {
			sortedArr.push(...bucket[i]);
		}
	}
	return sortedArr;
}

// radix sort
const radixSort = function (arr) {
  let maxDigits = Math.max(...arr).toString().length;
  let bucket = [];
  for (let i = 0; i <= 9; i += 1) {
    bucket.push([]);
  }
  for (let digit = 0; digit < maxDigits; digit += 1) {
    for (let i = 0; i < arr.length; i += 1) {
      let num = arr[i];
      let placeValue = Math.pow(10, digit);
      let digitValue = Math.floor(num / placeValue) % 10;
      bucket[digitValue].push(num);
    }
    let sortedArr = [];
    for (let i = 0; i < bucket.length; i += 1) {
      if (bucket[i].length !== 0) sortedArr.push(...bucket[i]);
    }
    for (let i = 0; i <= 9; i += 1) {
      bucket[i] = [];
    }
    arr = sortedArr;
  }
  return arr;
}

// two way merge
const merge = function (arr1, arr2) {
	let mergedArr = [];
	let i = 0;
	let j = 0;
	while (i < arr1.length && j < arr2.length) {
		if (arr1[i] < arr2[j]) {
			mergedArr.push(arr1[i]);
			i += 1;
		} else {
			mergedArr.push(arr2[j]);
			j += 1;
		}
	}
	while (i < arr1.length) {
		mergedArr.push(arr1[i]);
		i += 1;
	}
	while (j < arr2.length) {
		mergedArr.push(arr2[j]);
		j += 1;
	}
	return mergedArr;
}

// merging within one array
const mergeWithinOne = function (arr, low, mid, high) {
	let i = low;
	let j = mid+1;
	let mergeArr = [];
	while (i <= mid && j <= high) {
		if (arr[i] < arr[j]) {
			mergeArr.push(arr[i]);
			i += 1;
		} else {
			mergeArr.push(arr[j]);
			j += 1;
		}
	}
	while (i <= mid) {
		mergeArr.push(arr[i]);
		i += 1;
	}
	while (j <= high) {
		mergeArr.push(arr[j]);
		j += 1;
	}
	for (let k = 0; k < mergeArr.length; k += 1) {
		arr[k+low] = mergeArr[k];
	}
} 

// merge sort iterative version
const mergeSortI = function (arr) {
	let p = 2;
	for (; p <= arr.length; p *= 2) {
		for (let i = 0; i + p <= arr.length; i += p) {
			let low = i;
			let high = i + p - 1;
			let mid = Math.floor((high + low) / 2);
			mergeWithinOne(arr, low, mid, high);
		}
	}
	if (p/2 < arr.length) {
		mergeWithinOne(arr, 0, p/2, arr.length-1);
	}
}

// merge sort recursive version
const mergeSortR = function (arr, low, high) {
 	if (low < high) {
 		let mid = Math.floor((high+low)/2);
 		mergeSortR(arr, low, mid);
 		mergeSortR(arr, mid+1, high);
 		mergeWithinOne(arr, low, mid, high);
 	}
}

// sorting the pivot to the right position
const partition = function (arr, low, high) {
	let i = low;
	let j = high;
	let pivot = i;
	while (i <= j) {
		while (arr[i] < arr[pivot]) i += 1;
		while (arr[j] > arr[pivot]) j -= 1;
		if (i <= j) {
			let temp = arr[i];
			arr[i] = arr[j];
			arr[j] = temp;
			i += 1;
			j -= 1;
		}
	}
	let temp = arr[pivot];
	arr[pivot] = arr[j];
	arr[j] = temp;
	return j;
}

// quick sort
const quickSort = function (arr, low, high) {
	if (low < high) {
		let pivot = partition(arr, low, high);
		quickSort(arr, low, pivot-1);
		quickSort(arr, pivot+1, high);
	}
}
*/

File: sorting/sorting.py
import math

def bubble_sort(arr: list) -> None:
	'''
	bubble sort
	'''
	for i in range(len(arr)):
		for j in range(i, len(arr)):
			if arr[i] > arr[j]:
				temp = arr[i]
				arr[i] = arr[j]
				arr[j] = temp

def selection_sort(arr: list) -> None:
	'''
	selection sort
	'''
	for i in range(len(arr)):
		min = i 
		for j in range(i + 1, len(arr)):
			if arr[j] < arr[min]:
				min = j
		temp = arr[i]
		arr[i] = arr[min]
		arr[min] = temp

def insertion_sort(arr: list) -> None:
	'''
	insertion sort
	'''
	for i in range(1, len(arr)):
		min_index = i
		min_value = arr[i]
		for j in range(i - 1, -1, -1):
			if arr[j] > min_value:
				arr[j+1] = arr[j]
				min_index = j
		arr[min_index] = min_value

def count_sort(arr: list) -> list:
	'''
	count sort
	'''
	max_value = max(arr)
	bucket = [0] * (max_value + 1)
	for i in range(len(bucket)):
		bucket[arr[i]] += 1
	sorted_arr = list()
	i = 0
	while i < len(bucket):
		if bucket[i] > 0:
			sorted_arr.append(i)
			bucket[i] -= 1
		else:
			i += 1
	return sorted_arr

def bucket_sort(arr: list) -> list:
	'''
	bucket sort
	'''
	bucket = list()
	for i in range(10):
		bucket.append(list())
	for i in arr:
		bucket[int(i / 10)].append(i)
	for i in range(len(bucket)):
		bucket[i].sort()
	sorted_arr = list()
	for i in range(len(bucket)):
		if len(bucket[i]) > 0:
			sorted_arr.append(bucket[i].copy())
	return sorted_arr

def radix_sort(arr: list) -> list:
    '''
    radix sort
    '''
    max_digits = len(str(max(arr)))
    bucket = [[] for _ in range(10)]
    for digit in range(max_digits):
        for i in range(len(arr)):
            num = arr[i]
            place_value = int(math.pow(10, digit))
            digit_value = math.floor(num / place_value) % 10
            bucket[digit_value].append(num)
        sorted_arr = []
        for i in range(len(bucket)):
            if len(bucket[i]) != 0:
                sorted_arr.extend(bucket[i])
            bucket[i] = []
        arr = sorted_arr
    return arr


def two_way_merge(arr1: list, arr2: list) -> list:
	'''
	two way merging
	'''
	arr1 = sorted(arr1)
	arr2 = sorted(arr2)
	i = 0
	j = 0
	merged_arr = list()
	while (i < len(arr1) and j < len(arr2)):
		if arr1[i] < arr2[j]:
			merged_arr.append(arr1[i])
			i += 1
		else:
			merged_arr.append(arr2[j])
			j += 1
	while i < len(arr1):
		merged_arr.append(arr1[i])
		i += 1
	while j < len(arr2):
		merged_arr.append(arr2[j])
		j += 1
	return merged_arr

def merge_within_one(arr: list, low: int, mid: int, high: int) -> None:
	'''
	merging within one array
	'''
	i = low 
	j = mid+1
	merged_arr = list()
	while i <= mid and j <= high:
		if arr[i] < arr[j]:
			merged_arr.append(arr[i])
			i += 1
		else:
			merged_arr.append(arr[j])
			j += 1
	while i <= mid:
		merged_arr.append(arr[i])
		i += 1
	while j <= high:
		merged_arr.append(arr[j])
		j += 1
	for i in range(len(merged_arr)):
		arr[i+low] = merged_arr[i]

def merge_sort_i(arr: list) -> None:
	'''
	merge sort iterative version
	'''
	p = 2
	while p <= len(arr):
		i = 0
		while i + p <= len(arr):
			low = i
			high = i + p - 1
			mid = math.floor((low + high)/2)
			merge_within_one(arr, low, mid, high)
			i += p 
		p *= 2
	if p / 2 < len(arr):
		merge_within_one(arr, 0, int(p/2), len(arr)-1)

def merge_sort_r(arr: list, low: int, high: int) -> None:
	'''
	merge sort recursive version
	'''
	if low < high:
		mid = math.floor((low + high)/2)
		merge_sort_r(arr, low, mid)
		merge_sort_r(arr, mid+1, high)
		merge_within_one(arr, low, mid, high)


def partition(arr: list, low: int, high: int) -> int:
	'''
	sorting the pivot to the right position
	'''
	i = low 
	j = high 
	pivot = i 
	while i <= j:
		while arr[i] < arr[pivot]:
			i += 1
		while arr[j] > arr[pivot]:
			j -= 1
		if i <= j:
			temp = arr[i]
			arr[i] = arr[j]
			arr[j] = temp
			i += 1
			j -= 1
	temp = arr[pivot]
	arr[pivot] = arr[j]
	arr[j] = temp
	return j

def quick_sort(arr: list, low: int, high: int) -> None:
	'''
	quick sort
	'''
	if low < high:
		pivot = partition(arr, low, high)
		quick_sort(arr, low, pivot-1)
		quick_sort(arr, pivot+1, high)